lib/faktory/testing.rb

```diff
@@ -0,0 +1,329 @@
+require 'faktory/testing/runner'
+
+module Faktory
+  module Testing
+    def self.__test_mode
+      @__test_mode
+    end
+
+    def self.__test_mode=(mode)
+      @__test_mode = mode
+    end
+
+    def self.__set_test_mode(mode)
+      if block_given?
+        current_mode = self.__test_mode
+        begin
+          self.__test_mode = mode
+          yield
+        ensure
+          self.__test_mode = current_mode
+        end
+      else
+        self.__test_mode = mode
+      end
+    end
+
+    def self.disable!(&block)
+      __set_test_mode(:disable, &block)
+    end
+
+    def self.fake!(&block)
+      __set_test_mode(:fake, &block)
+    end
+
+    def self.inline!(&block)
+      # Don't allow blockless inline
+      # https://github.com/mperham/sidekiq/issues/3495
+      unless block_given?
+        raise 'Must provide a block to Faktory::Testing.inline!'
+      end
+
+      __set_test_mode(:inline, &block)
+    end
+
+    def self.enabled?
+      self.__test_mode != :disable
+    end
+
+    def self.disabled?
+      self.__test_mode == :disable
+    end
+
+    def self.fake?
+      self.__test_mode == :fake
+    end
+
+    def self.inline?
+      self.__test_mode == :inline
+    end
+
+    def self.constantize(str)
+      names = str.split('::')
+      names.shift if names.empty? || names.first.empty?
+
+      names.inject(Object) do |constant, name|
+        constant.const_defined?(name) ? constant.const_get(name) : constant.const_missing(name)
+      end
+    end
+  end
+
+  # Fake it 'til you make it by default
+  Faktory::Testing.fake!
+
+  class EmptyQueueError < RuntimeError; end
+
+  class Client
+    alias_method :real_push, :push
+    alias_method :real_open, :open
+
+    def push(job)
+      if Faktory::Testing.inline?
+        Faktory::Testing::Runner.new(job).push
```

Comment:
Doesn't look like this results in a JSON serialization/deserialization roundtrip.  That's necessary to get argument marshalling and other possible side effects.

Reply:
See right below L85-86 for an example.

Reply:
The Runner class is only used here and does almost nothing.  Pull the code into this block and remove the abstraction.